File: app/src/main/java/vn/ngh/stickyitem/decoration/StickyDecoration.java
package vn.ngh.stickyitem.decoration;

import android.graphics.Canvas;
import android.graphics.Rect;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.view.View;
import android.view.ViewGroup;

import vn.ngh.stickyitem.interfaces.StickyAdapter;

/**
 * Created by Ngh on 7/13/2017.
 */

public class StickyDecoration extends RecyclerView.ItemDecoration {
    private ViewHolder mViewHolderCache;
    private StickyAdapter mAdapter;
    private Rect mRect;
    private OnPositionListener mOnPositionListener;

    public StickyDecoration(StickyAdapter adapter) {
        mAdapter = adapter;
        mViewHolderCache = null;
        mRect = new Rect();
    }

    public void setOnPostitionListener(OnPositionListener event) {
        this.mOnPositionListener = event;
    }

    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
        int position = parent.getChildAdapterPosition(view);
        int headerHeight = 0;
        if (position != RecyclerView.NO_POSITION && mAdapter.getStickyIndex() >= 0) {
            View header = getSticky(parent, position).itemView;
            headerHeight = header.getHeight();
            if (position == mAdapter.getStickyIndex()) {
                outRect.set(0, headerHeight, 0, 0);
                return;
            }
            if (position >= mAdapter.getDataCount() - 1 && mCurrentPosition == Position.BOTTOM) {
                outRect.set(0, 0, 0, headerHeight);
                return;
            }
        }
        outRect.set(0, 0, 0, 0);
    }

    public StickyAdapter getAdapter() {
        return mAdapter;
    }

    public Rect getStickyRect() {
        return mRect;
    }

    public ViewHolder getViewHolder() {
        return mViewHolderCache;
    }

    private RecyclerView.ViewHolder getSticky(RecyclerView parent, int position) {
        int key = mAdapter.getStickyIndex();

        if (key >= 0) {
            if (!mAdapter.getEnableUpdate()) {
                return mViewHolderCache;
            } else {
                RecyclerView.ViewHolder holder = mAdapter.onCreateStickyViewHolder(parent);
                View sticky = holder.itemView;

                mAdapter.onBindStickyViewHolder(holder, position);
                int widthSpec = View.MeasureSpec.makeMeasureSpec(parent.getWidth(), View.MeasureSpec.EXACTLY);
                int heightSpec = View.MeasureSpec.makeMeasureSpec(parent.getHeight(), View.MeasureSpec.UNSPECIFIED);

                int childWidth = ViewGroup.getChildMeasureSpec(widthSpec,
                        parent.getPaddingLeft() + parent.getPaddingRight(), sticky.getLayoutParams().width);
                int childHeight = ViewGroup.getChildMeasureSpec(heightSpec,
                        parent.getPaddingTop() + parent.getPaddingBottom(), sticky.getLayoutParams().height);

                sticky.measure(childWidth, childHeight);
                sticky.layout(0, 0, sticky.getMeasuredWidth(), sticky.getMeasuredHeight());

                // save cache
                mViewHolderCache = holder;

                // update done so don't need create view any more
                mAdapter.setEnableUpdate(false);

                return holder;
            }
        }
        return null;
    }

    private Position mCurrentPosition = Position.BOTTOM;

    @Override
    public void onDrawOver(Canvas c, RecyclerView parent, RecyclerView.State state) {
        final int count = parent.getChildCount();

        int lastVisibleItemPosition =
                ((LinearLayoutManager) parent.getLayoutManager()).findLastVisibleItemPosition();
        int firstVisiblePosition =
                ((LinearLayoutManager) parent.getLayoutManager())
                        .findFirstCompletelyVisibleItemPosition();
        int paddingTop = parent.getPaddingTop();
        int paddingBottom = parent.getPaddingBottom();

        int indexSticky = mAdapter.getStickyIndex();

        if (indexSticky >= 0) {
            for (int layoutPos = 0; layoutPos < count; layoutPos++) {
                final View child = parent.getChildAt(layoutPos);

                final int adapterPos = parent.getChildAdapterPosition(child);

                if (adapterPos != RecyclerView.NO_POSITION) {
                    View header = getSticky(parent, adapterPos).itemView;
                    //TODO: find view to click (finding another way)
                    //View reset = header.findViewById(R.id.button_reset_or_luck_rank);
                    c.save();
                    int left = child.getLeft();
                    int top = 0;
                    if (firstVisiblePosition <= indexSticky && indexSticky <= lastVisibleItemPosition) {
                        if (adapterPos == indexSticky) {
                            top = getStickyTop(parent, child);
                            setPosition(Position.IN_LIST);
                        } else {
                            continue;
                        }
                    } else if (firstVisiblePosition > indexSticky) {
                        top = paddingTop;
                        setPosition(Position.TOP);
                    } else {
                        setPosition(Position.BOTTOM);
                        top = parent.getHeight() - header.getHeight() - paddingBottom;
                    }
                    c.translate(left, top);
                    header.draw(c);
                    c.restore();
                    // find location
//                    mRect.set(header.getWidth() - reset.getWidth(), top, header.getRight(),
//                            top + header.getHeight() - reset.getHeight());
                }
            }
        }
    }

    private void setPosition(Position position) {
        mCurrentPosition = position;
        if (mOnPositionListener != null) {
            mOnPositionListener.onSticky(position);
        }
    }

    private int getStickyTop(RecyclerView parent, View child) {
        RecyclerView.LayoutManager layoutManager = parent.getLayoutManager();

        int paddingTop = parent.getPaddingTop();
        int paddingBottom = parent.getPaddingBottom();
        int maxBottom = parent.getHeight() - child.getHeight() - paddingBottom - paddingTop;
        int pos = layoutManager.getDecoratedTop(child);
        int result = 0;

        if (pos < 0) {
            result = 0;
        } else if (pos > maxBottom) {
            result = maxBottom;
        } else {
            result = pos;
        }
        return result;
    }

    public void clearStickyCache() {
        mViewHolderCache = null;
    }

    public interface OnPositionListener {
        void onSticky(Position pos);
    }

    public enum Position {
        TOP, BOTTOM, IN_LIST
    }
}

File: app/src/main/java/vn/ngh/stickyitem/decoration/StickyDecoration.kt
package vn.ngh.stickyitem.decoration

import android.graphics.Canvas
import android.graphics.Rect
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.ViewHolder
import android.view.View
import android.view.ViewGroup

import vn.ngh.stickyitem.interfaces.StickyAdapter

/**
 * Created by Ngh
 */

open class StickyDecoration<T : ViewHolder>(private val adapter: StickyAdapter<T>) : RecyclerView.ItemDecoration() {
    private var viewHolder: ViewHolder? = null
    private val stickyRect: Rect
    private var mOnPositionListener: OnPositionListener? = null

    private var mCurrentPosition = Position.BOTTOM

    init {
        viewHolder = null
        stickyRect = Rect()
    }

    fun setOnPostitionListener(event: OnPositionListener) {
        this.mOnPositionListener = event
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State?) {
        val position = parent.getChildAdapterPosition(view)
        var headerHeight = 0
        if (position != RecyclerView.NO_POSITION && adapter.getStickyIndex() >= 0) {
            val header = getSticky(parent, position)?.itemView
            header?.let {
                headerHeight = it.height
                if (position == adapter.getStickyIndex()) {
                    outRect.set(0, headerHeight, 0, 0)
                    return
                }
                if (position >= adapter.getDataCount() - 1 && mCurrentPosition == Position.BOTTOM) {
                    outRect.set(0, 0, 0, headerHeight)
                    return
                }
            }
        }
        outRect.set(0, 0, 0, 0)
    }

    private fun getSticky(parent: RecyclerView, position: Int): RecyclerView.ViewHolder? {
        val key = adapter.getStickyIndex()

        if (key >= 0) {
            if (!adapter.getEnableUpdate()) {
                return viewHolder
            } else {
                val holder = adapter.onCreateStickyViewHolder(parent)
                val sticky = holder.itemView

                adapter.onBindStickyViewHolder(holder, position)
                val widthSpec = View.MeasureSpec.makeMeasureSpec(parent.width, View.MeasureSpec.EXACTLY)
                val heightSpec = View.MeasureSpec.makeMeasureSpec(parent.height, View.MeasureSpec.UNSPECIFIED)

                val childWidth = ViewGroup.getChildMeasureSpec(widthSpec,
                        parent.paddingLeft + parent.paddingRight, sticky.layoutParams.width)
                val childHeight = ViewGroup.getChildMeasureSpec(heightSpec,
                        parent.paddingTop + parent.paddingBottom, sticky.layoutParams.height)

                sticky.measure(childWidth, childHeight)
                sticky.layout(0, 0, sticky.measuredWidth, sticky.measuredHeight)

                // save cache
                viewHolder = holder

                // update done so don't need create view any more
                adapter.setEnableUpdate(false)

                return holder
            }
        }
        return null
    }

    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State?) {
        val count = parent.childCount

        val lastVisibleItemPosition = (parent.layoutManager as LinearLayoutManager).findLastVisibleItemPosition()
        val firstVisiblePosition = (parent.layoutManager as LinearLayoutManager)
                .findFirstCompletelyVisibleItemPosition()
        val paddingTop = parent.paddingTop
        val paddingBottom = parent.paddingBottom

        val indexSticky = adapter.getStickyIndex()

        if (indexSticky >= 0) {
            for (layoutPos in 0 until count) {
                val child = parent.getChildAt(layoutPos)

                val adapterPos = parent.getChildAdapterPosition(child)

                if (adapterPos != RecyclerView.NO_POSITION) {
                    val header = getSticky(parent, adapterPos)?.itemView
                    c.save()
                    val left = child.left
                    var top = 0
                    if (indexSticky in firstVisiblePosition..lastVisibleItemPosition) {
                        if (adapterPos == indexSticky) {
                            top = getStickyTop(parent, child)
                            setPosition(Position.IN_LIST)
                        } else {
                            continue
                        }
                    } else if (firstVisiblePosition > indexSticky) {
                        top = paddingTop
                        setPosition(Position.TOP)
                    } else {
                        setPosition(Position.BOTTOM)
                        if (header != null) {
                            top = parent.height - header.height - paddingBottom
                        }
                    }
                    c.translate(left.toFloat(), top.toFloat())
                    header?.draw(c)
                    c.restore()
                }
            }
        }
    }

    private fun setPosition(position: Position) {
        mCurrentPosition = position
        mOnPositionListener?.onSticky(position)
    }

    private fun getStickyTop(parent: RecyclerView, child: View): Int {
        val layoutManager = parent.layoutManager

        val paddingTop = parent.paddingTop
        val paddingBottom = parent.paddingBottom
        val maxBottom = parent.height - child.height - paddingBottom - paddingTop
        val pos = layoutManager.getDecoratedTop(child)
        var result = 0

        if (pos < 0) {
            result = 0
        } else if (pos > maxBottom) {
            result = maxBottom
        } else {
            result = pos
        }
        return result
    }

    fun clearStickyCache() {
        viewHolder = null
    }

    interface OnPositionListener {
        fun onSticky(pos: Position)
    }

    enum class Position {
        TOP, BOTTOM, IN_LIST
    }
}
